mathics/eval/arithmetic.py

```diff
@@ -145,176 +155,213 @@ def append_last():
+                item
+            ):
+                exp = eval_Plus(Integer1, previous_elem.elements[1])
+                elements[-1] = Expression(
+                    SymbolPower,
+                    item,
+                    exp,
+                )
+                continue
         else:
-            number = sympy.Mul(*[item.to_sympy() for item in numbers])
-            number = from_sympy(number)
-    else:
-        number = Integer1
+            item = item
+        # Otherwise, just append the element...
+        elements.append(item)
+
+    number = eval_multiply_numbers(*numbers) if numbers else Integer1
+
+    if len(elements) == 0 or number is Integer0:
+        return number
 
-    if number.sameQ(Integer1):
-        number = None
-    elif number.is_zero:
-        if not infinity_factor:
-            return number
-        # else, they are handled using the DirectedInfinity upvalues rules.
-    elif number.sameQ(IntegerM1) and elements and elements[0].has_form("Plus", None):
+    if number is IntegerM1 and elements and elements[0].has_form("Plus", None):
         elements[0] = Expression(
             elements[0].get_head(),
             *[
                 Expression(SymbolTimes, IntegerM1, element)
                 for element in elements[0].elements
             ],
         )
-        number = None
+        number = Integer1
 
-    if number is not None:
+    if number is not Integer1:
         elements.insert(0, number)
 
-    if not elements:
-        # if infinity_factor:
-        #    return SymbolComplexInfinity
+    if len(elements) == 1:
+        return elements[0]
+
+    elements = sorted(elements)
+    items_elements = items
+    if len(elements) == len(items_elements) and all(
+        elem.sameQ(item) for elem, item in zip(elements, items_elements)
+    ):
+        return None
+    return Expression(
+        SymbolTimes,
+        *elements,
+        elements_properties=ElementsProperties(False, False, True),
+    )
+
+
+def eval_add_numbers(
+    *numbers: Tuple[Number],
+) -> Number:
+    """
+    Add the elements in ``numbers``.
+    """
+    if len(numbers) == 0:
+        return Integer0
+    if len(numbers) == 1:
+        return numbers[0]
+
+    is_machine_precision = any(number.is_machine_precision() for number in numbers)
+    if is_machine_precision:
+        numbers = (item.to_mpmath() for item in numbers)
+        number = mpmath.fsum(numbers)
+        return from_mpmath(number)
+
+    prec = min_prec(*numbers)
+    if prec is not None:
+        # For a sum, what is relevant is the minimum accuracy of the terms
+        with mpmath.workprec(prec):
+            numbers = (item.to_mpmath() for item in numbers)
+            number = mpmath.fsum(numbers)
+            return from_mpmath(number, precision=prec)
+    else:
+        return from_sympy(sum(item.to_sympy() for item in numbers))
+
+
+def eval_multiply_numbers(*numbers: Tuple[Number]) -> Number:
```

Comment:
For now we should leave off a type annotation here. `Tuple[Number]` is not quite right.

Reply:
`-> Number` is not statically correct either, so for now let's leave it off.  We have no guarantee that the return of `from_sympy()` is indeed a Number. Statically it could be a Symbol.

I think we would need something like `assert instance(return_value, Number)` for static typing to accept this.